src/sb-util.ts

```diff
@@ -1,32 +1,215 @@
-import { Queryable, ScratchProjectOptions, AssetFetcher } from './abstracts';
-import { ProjectSource, Sb3Fetcher, ProjectJsonFetcher, ProjectByCloudIdFetcher } from './asset-fetcher';
+import { Queryable, AssetFetcher, SpriteProperties, SpritePosition, BlockProperties } from './abstracts';
+import { Sb3Fetcher, ProjectJsonFetcher, ProjectByCloudIdFetcher } from './asset-fetcher';
+
+enum CollectionTypes {
+	SPRITES = 'sprites',
+	BLOCKS = 'blocks',
+	ASSETS = 'assets'
+}
+
+enum ScratchProjectKeys {
+	TARGETS = 'targets'
+}
+
+enum SelectorSyntax {
+	OPEN_BRACKET = '[',
+	CLOSED_BRACKET = ']',
+	EQUALS = '=',
+	DOUBLE_QUOTE = '\"',
+	SINGLE_QUOTE = '\''
+}
+
+enum SpriteAttributes {
+	BLOCKS = 'blocks',
+	BROADCASTS = 'broadcasts',
+	LISTS = 'lists'
+}
+
+enum BlockAttributes {
+	ID = 'id'
+}
+
+/*
+sb-util CLASSES.
+
+The classes below make up the sb-util API.
+*/
+
+// This is the global storage for project data
+const storage = new WeakMap();
 
 export class ScratchProject implements Queryable {
 	constructor(projectJSON: Object, assetFetcher?: AssetFetcher) {
+		storage.set(this, projectJSON);
+	}
+
+	prop(property: string) {
+		return storage.get(this)[property];
 	}
 
-	query(selector?: String) {
-		return [];
+	query(selector: string) {
+		if(selector === CollectionTypes.SPRITES) {
+			// The key for sprites in the project JSON is called 'targets'
+			return this.prop(ScratchProjectKeys.TARGETS);
+		}
+	}
+
+	sprites(selector?: string){
+		const sprites = this.query(CollectionTypes.SPRITES);
+
+		if (!selector && typeof selector !== 'string') return new SpriteCollection(sprites);
+
+		return new SpriteCollection(sprites).query(selector);
+	}
+
+	stage() {
+		return this.sprites('[isStage=true]');
+	}
+
+	blocks() {
+		let blocks = [];
+		let sprites = this.sprites();
+		for (let s of sprites) {
+			const blocksObj = s.prop(SpriteAttributes.BLOCKS);
+			Object.entries(blocksObj).forEach(([key, val]) => {
+				let block = val;
+				block[BlockAttributes.ID] = key;
+				blocks.push(block);
+			})
```

Comment:
We need to setup linting. Missing semicolon.

Reply:
#10